File: ProgrammingChallenge/app/src/main/java/com/gabrielbmoro/programmingchallenge/domain/usecase/TopRatedMoviesUseCase.kt
package com.gabrielbmoro.programmingchallenge.domain.usecase

import com.gabrielbmoro.programmingchallenge.domain.model.Page
import com.gabrielbmoro.programmingchallenge.domain.usecase.mapper.MoviesMapper
import com.gabrielbmoro.programmingchallenge.repository.MoviesRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class TopRatedMoviesUseCase(val repository: MoviesRepository) {

    suspend fun execute(pageNumber: Int): Page? {
        return withContext(Dispatchers.IO){
            MoviesMapper.mapToPage(repository.getTopRatedMovies(pageNumber))
        }
    }
}

File: flutter-version/ProgrammingChallenge/movie_db_app/lib/domain/usecase/top_rated_movies_use_case.dart
import 'package:movie_db_app/domain/model/page.dart';
import 'package:movie_db_app/repository/movie_db_api_repository.dart';

class GetTopRatedMoviesUseCase {
  MovieDBRepository _repository;

  // ignore: non_constant_identifier_names
  GetTopRatedMoviesUseCase(MovieDBRepository repository) {
    _repository = repository;
  }

  Future<Page> execute(int pageNumber) {
    return _repository.getTopRatedMovies(
      pageNumber: pageNumber,
    );
  }
}
